File: solution_erg2/JAVA/Lab2.java

   
public class Lab2 {

    public static void main(String[] args) {

        int numThreads = 4; // local variable (in main) | not passed
		
		int n = 0; // local variable (in main) | pass-by-value
	    int[] a = new int[numThreads]; // local variable (in main) | pass-by-reference
	    for (int i = 0; i < numThreads; i++)
			a[i] = 0; 
		
		CounterThread counterThreads[] = new CounterThread[numThreads]; // local variable (in main) | not passed
		for (int i = 0; i < numThreads; i++) { // local variable (in main) | pass-by-value
		    counterThreads[i] = new CounterThread(i, n, a); // i (στοίχιση), n: pass-by-value arguments, a: pass-by-reference argument
		    counterThreads[i].start();
		}

        for (int i = 0; i < numThreads; i++) {
            try {
				counterThreads[i].join();
            }
            catch (InterruptedException e) {}
		}
		
		System.out.println("n = "+n);
		
		for (int i = 0; i < numThreads; i++)
			System.out.println("a["+i+"] = "+a[i]); 	
    }

}

class CounterThread extends Thread {
  	
    int threadID; // instance variable
    int n; // instance variable
    int[] a; // instance variable
    
    public CounterThread(int tid, int n, int[] a) { // tid, n: value parameters, a: reference paramter
        this.threadID = tid;
        this.n = n;
        this.a = a;
    }
  	
    public void run() {
        
        n = n + threadID ;
        a[threadID] = a[threadID] + n ;
        System.out.println("Thread "+threadID+ " my a = "+a[threadID]); 
    }
}


/* Lab2 is more straightforward than Lab1, mostly because we don't have any global variables.
 * The i and n arguments are passed by value, so n won't change and will remain 0 for the
 * duration of the processes runtime (also because it isn't returned by e.g. run). On the
 * other hand, a is obviously passed by reference. So every index will get a new value by the end.
 * Inside every thread, n will just become the value of threadId, and since a[0..3] = 0,
 * the value of every index of the array will essentially have the threadID number.
 * So the general output will be:
 * 
 * Thread i my a = i
 * n = 0
 * a[0] = 0
 * a[1] = 1
 * a[2] = 2
 * a[3] = 3
 * 
 * which proves to be true. Example output:
 * Thread 3 my a = 3
 * Thread 0 my a = 0
 * Thread 1 my a = 1
 * Thread 2 my a = 2
 * n = 0
 * a[0] = 0
 * a[1] = 1
 * a[2] = 2
 * a[3] = 3
 */

File: solution_erg2/C/Lab2.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>

#define NTHREADS 4

typedef struct
{
	int threadID; // struct member
	int n;        // struct member
	int *a;       // struct member
} thread_data;

void *counter_thread(void *ptr) // reference parameter
{
	int tID; // thread function var
	int n;   // thread function var
	int *a;  // thread function var | pointer to struct member a

	thread_data *thread_args = (thread_data*) ptr;
	tID = thread_args->threadID; // value parameter
	n   = thread_args->n; // value parameter
	a   = thread_args->a; // reference parameter

	n += tID;
	a[tID] = a[tID] + n;
	printf("Thread %d my a = %d\n", tID, a[tID]);
}

int main()
{
	int i; // main local var | pass-by-value
	int n; // main local var | pass-by-value
	int a[NTHREADS]; // main local var | pass-by-reference
        pthread_t threads[NTHREADS]; // main local var | not passed
	thread_data thread_args[NTHREADS]; // main local var | pass-by-reference

	n = 0;
	for (i = 0; i < NTHREADS; i++)
		a[i] = 0;

	for (i = 0; i < NTHREADS; i++)
	{
		thread_args[i].threadID = i;
		thread_args[i].n = n;
		thread_args[i].a = a;
		pthread_create(&threads[i], NULL, counter_thread, (void*) &thread_args[i]);
	}

	for (i = 0; i < NTHREADS; i++)
		pthread_join(threads[i], NULL);

	printf("n = %d\n", n);
	for (i = 0; i < NTHREADS; i++)
		printf("a[%d] = %d\n", i, a[i]);

        return 0;
}
